src/neo/Network/P2P/Payloads/Transaction.cs

```diff
@@ -217,7 +244,12 @@ public override int GetHashCode()
 
         public UInt160[] GetScriptHashesForVerifying(StoreView snapshot)
         {
-            var hashes = new HashSet<UInt160>(Cosigners.Keys) { Sender };
+            var hashes = new HashSet<UInt160>(Cosigners.Keys)
+            {
+                IsOracleResponse
+                ? Blockchain.GetConsensusAddress(NativeContract.Oracle.GetOracleNodes(snapshot))
+                : Sender
```

Comment:
We can add a check in `CheckOracleResponse` to replace here? It's no good to use `IsOracleResponse` in so many place.

Reply:
I may got your point, it's a so special oracle response tranaction! It doesn't feel very good because it destroys a lot of the nature of the tx itself.

Reply:
Any better idea?

Reply:
I don't like this way, the act of replacing the sender is too rude.This gives the platform excessive authority.

Reply:
The actor of replacing the sender is not good, giving the platform too much power.

Reply:
But we have limited the script of the response transaction, it is fixed. So the platform has no power.

https://github.com/neo-project/neo/blob/63b3a759f178a6c30d42844553b85f397c8376ad/src/neo/Network/P2P/Payloads/Transaction.cs#L219-L220

Reply:
Since we have `postPersist`, when user send `request` tx, which can pay a little more fee for the response tx. Then we can refund the remaining charges in `postPersist` method.

Reply:
It may create a new request in the callback.

Reply:
But if the response transaction consumes a lot of GAS, which may not what the sender expects, a large fee is deducted without any prior notice.

Reply:
Any idea?

Reply:
Can refer to my previous design.We can let the user directly pay a prepayment when requested, and then after the callback is executed, the remaining money will be refunded to the user。

> It may create a new request in the callback

It use "transfer" method to pay ,no effect on refund

Reply:
Transfer the fee to the Oracle contract? I don't think it's a good idea. Because the user's fees are mixed together, the cost of a response transaction can exceed the user's fee balance before you have a way to check it.

Reply:
I think that the existing payment behavior is more dangerous. Although the user will pay for the execution of the callback, this payment is not a completely controllable behavior.It is very likely that some errors or negligence caused the user to pay excess fees.

Reply:
We can put the sender of the request into the cosigner without changing the sender of the response, and at the same time limit the transfer permission of the Oracle address

Reply:
If this problem cannot be solved, I kind of want to go back to the synchronous solution.

Reply:
It can be solved. Transfer `reponseFee` to oracle contract, which can bind with request id, then the user's fees will not be mixed. It's kind of like prepayment.

Reply:
> the cost of a response transaction can exceed the user's fee balance before you have a way to check it.

When we verify response transaction, we can also check the balance in oracle contract by request id. And if the prepaid amount is exceeded, oracle node should send an error response transaction.

Reply:
If you check it in verification, you will not be able to save any state. This means that any number of transactions that do not exceed the prepaid amount will be passed, even if the sum of their fees is greater than the prepaid amount.

Reply:
We don't need to save any state. For each URL request, there is a corresponding prepaid amount for response tx(_Suppose one url request bind with one repsonse tx_). We just check whether the prepaid amount  is enough for the related response tx. 

If there are two different response txs with the same request id, it's another issue.

Reply:
I don't think this will happen.Because the user has already sent GAS to the Oracle contract every time a request is registered.If a request can be successfully registered, it means that the user has paid an prepaid payment for it.The Oracle node will use the prepayment value as the maximum value when it listens to the request and returns the response, so no matter how many transactions there are, the fee must be <prepayment*number of transactions.

Reply:
But the callback can create new requests. Who send the fees for these new requests?

Reply:
Payer as a parameter？

Reply:
No effect, because this part of the cost will be included in the original execution cost

Reply:
No effect, because this part of the cost will be included in the original execution cost

Reply:
The detailed process is : registration a request-->>addGAS( )-->>Oracle address mint GAS

Reply:
The user needs to pay a GAS as the execution fee of the response when registering a request.This fee does not need to be transferred using transfer, but instead uses the addGAS method and the mint method to add to the Oracle contract.When the response is executed, the execution fee is directly used as the system fee and network fee for the transaction.If a new request needs to be defined in the callback function, then AddGAS is still used, which does not cause excess
For details, please refer to my previous design.

Reply:
Maybe this is feasible. However, if the call chain is too long, users should have a way to add GAS to prevent insufficient fee.

Reply:
We can use this method first, if necessary in the future, we can add some interface operations such as supplement

Reply:
@doubiliu Please take a look. 6f47395

Reply:
I think we should disallow the callback to create new request, because it can lead to reentry attacks, such as repeated transfers of users' money.

Reply:
How does the attacks work? I think callback is the continuation of request, it should be able to get the same permissions as request.

Reply:
If the request tx have called `transfer` method, and the response tx trigger the request again.

Reply:
But you can call `transfer` in the request transaction any number of times. So they are the same. The callback is the continuation of request.

Reply:
The different is that whether users know it.  The longer the callback path, the greater the uncertainty for the user. What's more, the follow-up transaction is not sent by the user himself.

Reply:
They are completely equivalent. Because the response script is not arbitrary, but hard-coded.